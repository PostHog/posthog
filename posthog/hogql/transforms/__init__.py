from .asterisk import expand_asterisks
from .lazy_tables import resolve_lazy_tables
