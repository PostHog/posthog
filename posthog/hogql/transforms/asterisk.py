from typing import List

from posthog.hogql import ast
from posthog.hogql.visitor import TraversingVisitor


def expand_asterisks(node: ast.Expr):
    AsteriskExpander().visit(node)


class AsteriskExpander(TraversingVisitor):
    def visit_select_query(self, node: ast.SelectQuery):
        super().visit_select_query(node)

        columns: List[ast.Expr] = []
        for column in node.select:
            if isinstance(column.ref, ast.AsteriskRef):
                asterisk = column.ref
                if isinstance(asterisk.table, ast.BaseTableRef):
                    table = asterisk.table.resolve_database_table()
                    database_fields = table.get_asterisk()
                    for key in database_fields.keys():
                        ref = ast.FieldRef(name=key, table=asterisk.table)
                        columns.append(ast.Field(chain=[key], ref=ref))
                        node.ref.columns[key] = ref
                elif isinstance(asterisk.table, ast.SelectQueryRef) or isinstance(
                    asterisk.table, ast.SelectQueryAliasRef
                ):
                    select = asterisk.table
                    while isinstance(select, ast.SelectQueryAliasRef):
                        select = select.ref
                    if isinstance(select, ast.SelectQueryRef):
                        for name in select.columns.keys():
                            ref = ast.FieldRef(name=name, table=asterisk.table)
                            columns.append(ast.Field(chain=[name], ref=ref))
                            node.ref.columns[name] = ref
                    else:
                        raise ValueError("Can't expand asterisk (*) on subquery")
                else:
                    raise ValueError(f"Can't expand asterisk (*) on a ref of type {type(asterisk.table).__name__}")

            else:
                columns.append(column)
        node.select = columns
