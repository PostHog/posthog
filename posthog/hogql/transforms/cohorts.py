from typing import Set

from posthog.hogql import ast
from posthog.hogql.context import HogQLContext
from posthog.hogql.errors import HogQLException
from posthog.hogql.visitor import CloningVisitor, TraversingVisitor, clone_expr


def resolve_cohort_subqueries(node: ast.Expr, context: HogQLContext = None) -> ast.Expr:
    from posthog.models import Cohort

    # find all cohorts
    cohort_finder = CohortFinder()
    cohort_finder.visit(node)

    if len(cohort_finder.cohorts) == 0:
        return node

    # fetch them
    cohorts = Cohort.objects.filter(id__in=cohort_finder.cohorts, team_id=context.team_id).values_list(
        "id", "is_static"
    )
    static_cohorts = set(id for id, is_static in cohorts if is_static)

    if len(cohort_finder.cohorts) != len(cohorts):
        missing_cohorts = [str(a) for a in (set(cohort_finder.cohorts) - set(id for id, _ in cohorts))]
        raise HogQLException(
            f"Could not find cohort{'s' if len(missing_cohorts) > 1 else ''}: {', '.join(missing_cohorts)}"
        )

    node = CohortSwapper(static_cohorts=static_cohorts, context=context).visit(node)
    return node


class CohortFinder(TraversingVisitor):
    def __init__(self):
        super().__init__()
        self.cohorts: Set[int] = set()

    def visit_compare_operation(self, node: ast.CompareOperation):
        if node.op == ast.CompareOperationOp.InCohort or node.op == ast.CompareOperationOp.NotInCohort:
            if not isinstance(node.right, ast.Constant) or not isinstance(node.right.value, int):
                raise HogQLException("Cohort id must be a constant integer")
            self.cohorts.add(node.right.value)


class CohortSwapper(CloningVisitor):
    def __init__(self, context: HogQLContext, static_cohorts: Set[int]):
        super().__init__(clear_types=False)
        self.static_cohorts = static_cohorts
        self.context = context

    def visit_compare_operation(self, node: ast.CompareOperation):
        from posthog.hogql.property import cohort_subquery
        from posthog.hogql.printer import prepare_ast_for_printing

        if node.op == ast.CompareOperationOp.InCohort or node.op == ast.CompareOperationOp.NotInCohort:
            if not isinstance(node.right, ast.Constant) or not isinstance(node.right.value, int):
                raise HogQLException("Cohort id must be a constant integer")
            cohort_id = node.right.value
            node = clone_expr(node)
            node.left = super().visit(node.left)
            node.op = (
                ast.CompareOperationOp.In
                if node.op == ast.CompareOperationOp.InCohort
                else ast.CompareOperationOp.NotIn
            )
            node.right = cohort_subquery(cohort_id, cohort_id in self.static_cohorts)
            node.right = prepare_ast_for_printing(node.right, context=self.context, dialect="clickhouse")
            return node
        return node
