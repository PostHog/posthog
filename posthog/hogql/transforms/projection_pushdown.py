from collections import defaultdict
from typing import cast

from posthog.hogql import ast
from posthog.hogql.base import _T_AST
from posthog.hogql.context import HogQLContext
from posthog.hogql.visitor import TraversingVisitor


class ProjectionPushdownOptimizer(TraversingVisitor):
    """
    Top-down projection pushdown optimizer that prunes unused asterisk-expanded columns from subqueries.

    Algorithm Overview:
    ──────────────────
    This optimizer works in a single top-down pass through the query tree:

    Phase 1 - Register: Map subquery types to AST nodes for demand tracking
    Phase 2 - Collect: Gather column demands from WHERE/GROUP BY/ORDER BY/etc
    Phase 3 - Propagate: For demanded columns, visit their source to propagate to child queries
    Phase 4 - Recurse: Visit child subqueries (repeat phases 1-4)
    Phase 5 - Prune: Remove unreferenced asterisk columns from this query
    """

    def __init__(self):
        super().__init__()
        self.demands: dict[int, set[str]] = defaultdict(set)
        self.subquery_map: dict[int, ast.SelectQuery | ast.SelectSetQuery] = {}

    def visit_select_query(self, node: ast.SelectQuery) -> ast.SelectQuery:
        # Phase 1: Register subqueries for demand tracking
        if node.select_from:
            self._register_subqueries(node.select_from)

        # Phase 2: Collect column demands from query clauses
        # Skip asterisk-expanded SELECT columns - their demands flow from parent queries
        for expr in node.select:
            if not self._is_from_asterisk(expr):
                self.visit(expr)

        if node.where:
            self.visit(node.where)
        if node.prewhere:
            self.visit(node.prewhere)
        if node.group_by:
            for expr in node.group_by:
                self.visit(expr)
        if node.having:
            self.visit(node.having)
        if node.order_by:
            for expr in node.order_by:
                self.visit(expr)
        if node.limit:
            self.visit(node.limit)
        if node.offset:
            self.visit(node.offset)

        if node.select_from:
            self._collect_join_constraint_column_demands(node.select_from)

        # Phase 3: Propagate parent demands down to child subqueries
        self._propagate_demands_to_children(node)

        # Phase 4: Recursively visit and optimize child subqueries
        if node.select_from:
            self.visit(node.select_from)

        # Phase 5: Prune unreferenced asterisk columns from this query
        self._prune_columns(node)

        return node

    def visit_select_set_query(self, node: ast.SelectSetQuery) -> ast.SelectSetQuery:
        """
        Handle UNION/INTERSECT/EXCEPT queries.

        All branches must have identical column structure, so we:
        1. Propagate parent demands to all branches uniformly
        2. Visit each branch to apply pruning
        """
        # Propagate parent demands to all branches uniformly
        self._propagate_demands_to_union_branches(node)

        # Visit each branch
        self.visit(node.initial_select_query)
        for set_node in node.subsequent_select_queries:
            self.visit(set_node.select_query)

        return node

    def _is_from_asterisk(self, expr: ast.Expr) -> bool:
        """Check if an expression was expanded from asterisk"""
        if isinstance(expr, ast.Alias):
            expr = expr.expr  # whoops - we want to take the Field from the Alias
        if isinstance(expr, ast.Field):
            return expr.from_asterisk
        return False

    def _propagate_demands_to_children(self, node: ast.SelectQuery) -> None:
        """
        Propagate parent demands to child subqueries.

        When a parent query demands a column from us, we need to visit that column's
        source expression to propagate the demand down to our child subqueries.
        """
        demanded_from_this = self.demands.get(id(node))
        if not demanded_from_this:
            return

        for col_name in demanded_from_this:
            for expr in node.select:
                if self._get_column_name(expr) == col_name:
                    self.visit(expr)
                    break

    def _propagate_demands_to_union_branches(self, node: ast.SelectSetQuery) -> None:
        """
        Propagate demands from parent to all UNION/INTERSECT/EXCEPT branches.

        Since all branches must have identical column structure, we propagate
        the same demands to all of them.
        """
        demanded_from_this = self.demands.get(id(node))
        if not demanded_from_this:
            return

        all_queries = [node.initial_select_query] + [sn.select_query for sn in node.subsequent_select_queries]

        for query in all_queries:
            if isinstance(query, ast.SelectQuery | ast.SelectSetQuery):
                self.demands[id(query)].update(demanded_from_this)

    def _register_subqueries(self, from_clause: ast.JoinExpr) -> None:
        """Register all subqueries in FROM clause before collecting demands"""
        if from_clause.type is None:
            return

        if isinstance(from_clause.table, ast.SelectQuery):
            self._register_subquery(from_clause.table, from_clause.type)
        elif isinstance(from_clause.table, ast.SelectSetQuery):
            self._register_union_subquery(from_clause.table, from_clause.type)

        if from_clause.next_join:
            self._register_subqueries(from_clause.next_join)

    def _register_subquery(
        self,
        subquery: ast.SelectQuery,
        type_annotation: ast.BaseTableType | ast.SelectSetQueryType | ast.SelectQueryType | ast.SelectQueryAliasType,
    ) -> None:
        """Map type to subquery node for demand tracking"""
        # Map both the type and the inner SelectQueryType if it's an alias
        self.subquery_map[id(type_annotation)] = subquery
        if isinstance(type_annotation, ast.SelectQueryAliasType) and type_annotation.select_query_type:
            self.subquery_map[id(type_annotation.select_query_type)] = subquery

    def _register_union_subquery(
        self,
        union_query: ast.SelectSetQuery,
        type_annotation: ast.BaseTableType | ast.SelectSetQueryType | ast.SelectQueryType | ast.SelectQueryAliasType,
    ) -> None:
        """Map type to union subquery node for demand tracking"""
        self.subquery_map[id(type_annotation)] = union_query

        # Also register with the inner SelectSetQueryType for both aliased and non-aliased cases
        if isinstance(type_annotation, ast.SelectQueryAliasType):
            # Aliased: register with the inner select_query_type
            if type_annotation.select_query_type:
                self.subquery_map[id(type_annotation.select_query_type)] = union_query
        elif isinstance(type_annotation, ast.SelectSetQueryType):
            # Non-aliased: the type_annotation IS the SelectSetQueryType, but we also need to
            # register with the first branch's type for field resolution
            if type_annotation.types:
                self.subquery_map[id(type_annotation.types[0])] = union_query

    def _get_subquery(self, table_type: ast.Type) -> ast.SelectQuery | ast.SelectSetQuery | None:
        """Retrieve subquery by type"""
        return self.subquery_map.get(id(table_type))

    def visit_field(self, node: ast.Field) -> ast.Field:
        """Record demand when field references subquery column"""
        if not isinstance(node.type, ast.FieldType):
            return node

        table_type = node.type.table_type

        if isinstance(table_type, ast.SelectQueryType | ast.SelectQueryAliasType | ast.SelectSetQueryType):
            subquery = self._get_subquery(table_type)
            if subquery:
                self.demands[id(subquery)].add(node.type.name)

        return node

    def _collect_join_constraint_column_demands(self, from_clause: ast.JoinExpr) -> None:
        """Collect demands from JOIN constraints"""
        if from_clause.constraint:
            self.visit(from_clause.constraint)

        if from_clause.next_join:
            self._collect_join_constraint_column_demands(from_clause.next_join)

    def _prune_columns(self, node: ast.SelectQuery) -> None:
        """Prune asterisk-expanded columns that aren't demanded"""
        demanded = self.demands.get(id(node))
        if not demanded:
            return

        pruned_select = []
        for expr in node.select:
            if not self._is_from_asterisk(expr):
                # Keep explicitly written columns
                pruned_select.append(expr)
            else:
                # Keep demanded asterisk columns
                col_name = self._get_column_name(expr)
                if col_name and col_name in demanded:
                    pruned_select.append(expr)

        if pruned_select:
            node.select = pruned_select

    def _get_column_name(self, expr: ast.Expr) -> str | None:
        """Extract column name from expression"""
        if isinstance(expr, ast.Field):
            return str(expr.chain[-1]) if expr.chain else None
        elif isinstance(expr, ast.Alias):
            return expr.alias
        return None


def pushdown_projections(node: _T_AST, context: HogQLContext) -> _T_AST:
    """Prune unused columns from asterisk expansions in subqueries"""
    optimizer = ProjectionPushdownOptimizer()
    return cast(_T_AST, optimizer.visit(node))
