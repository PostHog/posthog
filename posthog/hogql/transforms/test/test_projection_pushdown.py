import pytest
from posthog.test.base import BaseTest

from posthog.schema import HogQLQueryModifiers

from posthog.hogql import ast
from posthog.hogql.context import HogQLContext
from posthog.hogql.parser import parse_select
from posthog.hogql.printer import prepare_ast_for_printing
from posthog.hogql.transforms.projection_pushdown import pushdown_projections


class TestProjectionPushdown(BaseTest):
    maxDiff = None
    snapshot: object

    def _optimize(self, query_str: str):
        """Helper: parse, resolve, and optimize a query"""
        modifiers = HogQLQueryModifiers(optimizeProjections=False)  # Disable automatic optimization
        context = HogQLContext(team_id=self.team.pk, enable_select_queries=True, modifiers=modifiers)
        query = parse_select(query_str)
        prepared = prepare_ast_for_printing(query, context, dialect="hogql")
        assert prepared is not None
        optimized = pushdown_projections(prepared, context)
        return optimized

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_simple_pushdown(self):
        """SELECT event FROM (SELECT * FROM events)"""
        optimized = self._optimize("SELECT event FROM (SELECT * FROM events) AS sub")

        # Inner query should only select 'event'
        inner_query = optimized.select_from.table
        assert len(inner_query.select) == 1

        # The column might be aliased
        first_col = inner_query.select[0]
        if hasattr(first_col, "alias"):
            assert first_col.alias == "event"
        else:
            assert first_col.chain == ["event"]
        assert first_col.from_asterisk if isinstance(first_col, ast.Field) else first_col.expr.from_asterisk

        assert optimized.to_hogql() == self.snapshot

    def _col_name(self, col):
        """Helper to get column name for debugging"""
        if hasattr(col, "alias"):
            return col.alias
        if hasattr(col, "chain") and col.chain:
            return col.chain[-1]
        return "?"

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_preserves_used_columns(self):
        """Columns in WHERE/GROUP BY also needed"""
        optimized = self._optimize("""
            SELECT event, count()
            FROM (SELECT * FROM events) AS sub
            WHERE distinct_id = 'test'
            GROUP BY timestamp
        """)

        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert column_names >= {"event", "distinct_id", "timestamp"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_no_pushdown_without_asterisk(self):
        """Don't modify queries without asterisk"""
        optimized = self._optimize("SELECT event FROM (SELECT event, distinct_id FROM events) AS sub")

        inner_query = optimized.select_from.table
        assert len(inner_query.select) == 2

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_nested_pushdown(self):
        """Pushdown through multiple levels"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT * FROM (
                    SELECT * FROM events
                ) AS inner
            ) AS outer
        """)

        # Both inner queries should only have 'event'
        # Check outer subquery
        outer_query = optimized.select_from.table
        assert len(outer_query.select) == 1
        assert self._col_name(outer_query.select[0]) == "event"

        # Check inner subquery
        inner_query = outer_query.select_from.table
        assert len(inner_query.select) == 1
        assert self._col_name(inner_query.select[0]) == "event"

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_preserves_explicit_columns(self):
        """Don't prune columns user explicitly wrote"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT event, distinct_id, timestamp FROM events
            ) AS sub
        """)

        # Should keep all columns (none have from_asterisk=True)
        inner_query = optimized.select_from.table
        assert len(inner_query.select) == 3

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_prewhere_clause(self):
        """PREWHERE clause columns should be preserved"""
        optimized = self._optimize("""
            SELECT event
            FROM (SELECT * FROM events) AS sub
            PREWHERE distinct_id = 'test'
        """)

        # Should include both event and distinct_id
        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert "event" in column_names
        assert "distinct_id" in column_names

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_order_by_clause(self):
        """ORDER BY clause columns should be preserved"""
        optimized = self._optimize("""
            SELECT event
            FROM (SELECT * FROM events) AS sub
            ORDER BY timestamp DESC
        """)

        # Should include both event and timestamp
        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert "event" in column_names
        assert "timestamp" in column_names

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_join_constraint_columns(self):
        """Columns in JOIN conditions should be preserved"""
        optimized = self._optimize("""
            SELECT e.event, e.distinct_id
            FROM (SELECT * FROM events) AS e
            LEFT JOIN sessions ON sessions.id = e.`$session_id`
        """)

        # Should include event, distinct_id, AND $session_id (from JOIN)
        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert "event" in column_names
        assert "distinct_id" in column_names
        assert "$session_id" in column_names

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_join_constraint_in_subquery(self):
        """Subquery JOIN constraints should preserve columns"""
        optimized = self._optimize("""
            SELECT * FROM (
                SELECT e.event, e.distinct_id, sessions.id
                FROM (SELECT * FROM events) AS e
                LEFT JOIN sessions ON sessions.id = e.`$session_id`
            )
        """)

        assert optimized.select_from is not None
        nested_query = optimized.select_from.table

        assert nested_query.select_from is not None
        inner_query = nested_query.select_from.table

        # Should include event, distinct_id, AND $session_id (from JOIN)
        column_names = {self._col_name(col) for col in inner_query.select}
        assert "event" in column_names
        assert "distinct_id" in column_names
        assert "$session_id" in column_names, f"Got columns: {column_names}"

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_multiple_joins_with_mixed_columns(self):
        """Complex query with multiple JOINs and mixed SELECT/WHERE demands"""
        optimized = self._optimize("""
            SELECT e.event, e2.event as event2
            FROM (SELECT * FROM events) AS e
            LEFT JOIN (SELECT * FROM events) AS e2 ON e2.distinct_id = e.distinct_id
            WHERE e.timestamp > '2024-01-01'
            ORDER BY e.created_at
        """)

        # First events subquery should have: event, distinct_id (JOIN), timestamp (WHERE), created_at (ORDER BY)
        events_query = optimized.select_from.table
        events_cols = {self._col_name(col) for col in events_query.select}
        assert events_cols >= {"event", "distinct_id", "timestamp", "created_at"}

        # Second events subquery should have: event (SELECT) and distinct_id (JOIN constraint)
        events2_query = optimized.select_from.next_join.table
        events2_cols = {self._col_name(col) for col in events2_query.select}
        assert events2_cols >= {"event", "distinct_id"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_deeply_nested_with_multiple_demands(self):
        """4 levels deep with demands at each level"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT event, distinct_id FROM (
                    SELECT event, distinct_id, timestamp FROM (
                        SELECT * FROM events
                    ) AS l3
                    WHERE timestamp > '2024-01-01'
                ) AS l2
                WHERE distinct_id = 'user1'
            ) AS l1
        """)

        # Level 1: event, distinct_id
        l1 = optimized.select_from.table
        l1_cols = {self._col_name(col) for col in l1.select}
        assert l1_cols == {"event", "distinct_id"}

        # Level 2: event, distinct_id, timestamp
        l2 = l1.select_from.table
        l2_cols = {self._col_name(col) for col in l2.select}
        assert l2_cols == {"event", "distinct_id", "timestamp"}

        # Level 3: Only event, distinct_id, timestamp (pruned from asterisk)
        l3 = l2.select_from.table
        l3_cols = {self._col_name(col) for col in l3.select}
        assert l3_cols == {"event", "distinct_id", "timestamp"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_mixed_asterisk_and_explicit_columns(self):
        """Query with both SELECT *, col1, col2 pattern"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT distinct_id, *, timestamp FROM events
            ) AS sub
        """)

        # Should keep distinct_id and timestamp (explicit) plus event (demanded from asterisk)
        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert "event" in column_names
        assert "distinct_id" in column_names
        assert "timestamp" in column_names

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_group_by_and_having_demands(self):
        """Ensure GROUP BY and HAVING columns are preserved"""
        optimized = self._optimize("""
            SELECT event, count() as cnt
            FROM (SELECT * FROM events) AS sub
            GROUP BY event, distinct_id
            HAVING count() > 10 AND timestamp > '2024-01-01'
        """)

        # Should include event (SELECT + GROUP BY), distinct_id (GROUP BY), timestamp (HAVING)
        inner_query = optimized.select_from.table
        column_names = {self._col_name(col) for col in inner_query.select}
        assert column_names >= {"event", "distinct_id", "timestamp"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_subquery_in_join_with_demands(self):
        """Nested subqueries in both sides of JOIN"""
        optimized = self._optimize("""
            SELECT e1.event, e2.timestamp
            FROM (
                SELECT * FROM (SELECT * FROM events) AS inner_e1
            ) AS e1
            LEFT JOIN (
                SELECT * FROM (SELECT * FROM events) AS inner_e2
            ) AS e2 ON e2.distinct_id = e1.distinct_id
            WHERE e1.properties != '{}'
        """)

        # Outer e1 subquery
        outer_e1 = optimized.select_from.table
        outer_e1_cols = {self._col_name(col) for col in outer_e1.select}
        assert outer_e1_cols >= {"event", "distinct_id", "properties"}

        # Inner e1 subquery should have same demands propagated
        inner_e1 = outer_e1.select_from.table
        inner_e1_cols = {self._col_name(col) for col in inner_e1.select}
        assert inner_e1_cols >= {"event", "distinct_id", "properties"}

        # Outer e2 subquery
        outer_e2 = optimized.select_from.next_join.table
        outer_e2_cols = {self._col_name(col) for col in outer_e2.select}
        assert outer_e2_cols >= {"timestamp", "distinct_id"}

        # Inner e2 subquery
        inner_e2 = outer_e2.select_from.table
        inner_e2_cols = {self._col_name(col) for col in inner_e2.select}
        assert inner_e2_cols >= {"timestamp", "distinct_id"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_union_all_without_alias(self):
        """UNION ALL without subquery alias should still be pruned"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT * FROM events UNION ALL SELECT * FROM events
            )
        """)

        # The UNION subquery should be a SelectSetQuery
        union_query = optimized.select_from.table
        assert isinstance(union_query, ast.SelectSetQuery)

        # Both branches should only have 'event' column
        first_branch = union_query.initial_select_query
        assert isinstance(first_branch, ast.SelectQuery)
        first_cols = {self._col_name(col) for col in first_branch.select}
        assert first_cols == {"event"}, f"Expected only 'event' but got {first_cols}"

        second_branch = union_query.subsequent_select_queries[0].select_query
        assert isinstance(second_branch, ast.SelectQuery)
        second_cols = {self._col_name(col) for col in second_branch.select}
        assert second_cols == {"event"}, f"Expected only 'event' but got {second_cols}"

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_union_all_with_asterisk(self):
        """UNION ALL branches with asterisk should be pruned uniformly"""
        optimized = self._optimize("""
            SELECT event, distinct_id FROM (
                SELECT * FROM events WHERE event = 'click'
                UNION ALL
                SELECT * FROM events WHERE event = 'pageview'
            ) AS sub
        """)

        # The UNION subquery should be a SelectSetQuery
        union_query = optimized.select_from.table
        assert isinstance(union_query, ast.SelectSetQuery)

        # Both branches should have the same columns (event, distinct_id)
        first_branch = union_query.initial_select_query
        assert isinstance(first_branch, ast.SelectQuery)
        first_cols = {self._col_name(col) for col in first_branch.select}
        assert first_cols == {"event", "distinct_id"}

        second_branch = union_query.subsequent_select_queries[0].select_query
        assert isinstance(second_branch, ast.SelectQuery)
        second_cols = {self._col_name(col) for col in second_branch.select}
        assert second_cols == {"event", "distinct_id"}

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_union_all_multiple_branches(self):
        """UNION ALL with 3+ branches should all be pruned uniformly"""
        optimized = self._optimize("""
            SELECT event FROM (
                SELECT * FROM events WHERE event = 'click'
                UNION ALL
                SELECT * FROM events WHERE event = 'pageview'
                UNION ALL
                SELECT * FROM events WHERE event = 'submit'
            ) AS sub
        """)

        union_query = optimized.select_from.table
        assert isinstance(union_query, ast.SelectSetQuery)

        # All three branches should only have 'event' column
        all_branches = [union_query.initial_select_query] + [
            sn.select_query for sn in union_query.subsequent_select_queries
        ]

        for branch in all_branches:
            assert isinstance(branch, ast.SelectQuery)
            cols = {self._col_name(col) for col in branch.select}
            assert cols == {"event"}, f"Expected only 'event' but got {cols}"

        assert optimized.to_hogql() == self.snapshot

    @pytest.mark.usefixtures("unittest_snapshot")
    def test_cte_with_asterisk_pushdown(self):
        """CTEs with asterisk should have projection pushdown applied after CTE inlining"""
        optimized = self._optimize("""
            WITH events_cte AS (
                SELECT * FROM events
            )
            SELECT event, distinct_id FROM events_cte
            WHERE timestamp > '2024-01-01'
        """)

        # After CTE inlining, the CTE becomes a subquery
        # The subquery should only have the columns we actually need
        assert optimized.select_from is not None
        subquery = optimized.select_from.table
        column_names = {self._col_name(col) for col in subquery.select}
        assert column_names >= {"event", "distinct_id", "timestamp"}
        # Verify columns have from_asterisk marker
        assert any(
            col.from_asterisk if isinstance(col, ast.Field) else col.expr.from_asterisk for col in subquery.select
        )

        assert optimized.to_hogql() == self.snapshot
