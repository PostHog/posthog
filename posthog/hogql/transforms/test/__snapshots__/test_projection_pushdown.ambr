# serializer version: 1
# name: TestProjectionPushdown.test_cte_with_asterisk_pushdown
  "SELECT event, distinct_id FROM (SELECT event, timestamp, distinct_id FROM events) AS events_cte WHERE greater(timestamp, '2024-01-01')"
# ---
# name: TestProjectionPushdown.test_deeply_nested_with_multiple_demands
  "SELECT event FROM (SELECT event, distinct_id FROM (SELECT event, distinct_id, timestamp FROM (SELECT event, timestamp, distinct_id FROM events) AS l3 WHERE greater(timestamp, '2024-01-01')) AS l2 WHERE equals(distinct_id, 'user1')) AS l1"
# ---
# name: TestProjectionPushdown.test_group_by_and_having_demands
  "SELECT event, count() AS cnt FROM (SELECT event, timestamp, distinct_id FROM events) AS sub GROUP BY event, distinct_id HAVING and(greater(count(), 10), greater(timestamp, '2024-01-01'))"
# ---
# name: TestProjectionPushdown.test_join_constraint_columns
  'SELECT e.event, e.distinct_id FROM (SELECT event, distinct_id, $session_id FROM events) AS e LEFT JOIN sessions ON equals(sessions.id, e.$session_id)'
# ---
# name: TestProjectionPushdown.test_join_constraint_in_subquery
  'SELECT event, distinct_id, id FROM (SELECT e.event, e.distinct_id, sessions.id FROM (SELECT event, distinct_id, $session_id FROM events) AS e LEFT JOIN sessions ON equals(sessions.id, e.$session_id))'
# ---
# name: TestProjectionPushdown.test_mixed_asterisk_and_explicit_columns
  'SELECT event FROM (SELECT distinct_id, event, timestamp FROM events) AS sub'
# ---
# name: TestProjectionPushdown.test_multiple_joins_with_mixed_columns
  "SELECT e.event, e2.event AS event2 FROM (SELECT event, timestamp, distinct_id, created_at FROM events) AS e LEFT JOIN (SELECT event, distinct_id FROM events) AS e2 ON equals(e2.distinct_id, e.distinct_id) WHERE greater(e.timestamp, '2024-01-01') ORDER BY e.created_at ASC"
# ---
# name: TestProjectionPushdown.test_nested_pushdown
  'SELECT event FROM (SELECT event FROM (SELECT event FROM events) AS inner) AS outer'
# ---
# name: TestProjectionPushdown.test_no_pushdown_without_asterisk
  'SELECT event FROM (SELECT event, distinct_id FROM events) AS sub'
# ---
# name: TestProjectionPushdown.test_order_by_clause
  'SELECT event FROM (SELECT event, timestamp FROM events) AS sub ORDER BY timestamp DESC'
# ---
# name: TestProjectionPushdown.test_preserves_explicit_columns
  'SELECT event FROM (SELECT event, distinct_id, timestamp FROM events) AS sub'
# ---
# name: TestProjectionPushdown.test_preserves_used_columns
  "SELECT event, count() FROM (SELECT event, timestamp, distinct_id FROM events) AS sub WHERE equals(distinct_id, 'test') GROUP BY timestamp"
# ---
# name: TestProjectionPushdown.test_prewhere_clause
  "SELECT event FROM (SELECT event, distinct_id FROM events) AS sub PREWHERE equals(distinct_id, 'test')"
# ---
# name: TestProjectionPushdown.test_simple_pushdown
  'SELECT event FROM (SELECT event FROM events) AS sub'
# ---
# name: TestProjectionPushdown.test_subquery_in_join_with_demands
  "SELECT e1.event, e2.timestamp FROM (SELECT event, properties, distinct_id FROM (SELECT event, properties, distinct_id FROM events) AS inner_e1) AS e1 LEFT JOIN (SELECT timestamp, distinct_id FROM (SELECT timestamp, distinct_id FROM events) AS inner_e2) AS e2 ON equals(e2.distinct_id, e1.distinct_id) WHERE notEquals(e1.properties, '{}')"
# ---
# name: TestProjectionPushdown.test_union_all_multiple_branches
  "SELECT event FROM (SELECT event FROM events WHERE equals(event, 'click') UNION ALL SELECT event FROM events WHERE equals(event, 'pageview') UNION ALL SELECT event FROM events WHERE equals(event, 'submit')) AS sub"
# ---
# name: TestProjectionPushdown.test_union_all_with_asterisk
  "SELECT event, distinct_id FROM (SELECT event, distinct_id FROM events WHERE equals(event, 'click') UNION ALL SELECT event, distinct_id FROM events WHERE equals(event, 'pageview')) AS sub"
# ---
# name: TestProjectionPushdown.test_union_all_without_alias
  'SELECT event FROM (SELECT event FROM events UNION ALL SELECT event FROM events)'
# ---
